onchain/rollups/contracts/consensus/authority/AuthorityHistoryPairFactory.sol

```diff
@@ -0,0 +1,112 @@
+contract AuthorityHistoryPairFactory is IAuthorityHistoryPairFactory {
+    IAuthorityFactory immutable authorityFactory;
+    IHistoryFactory immutable historyFactory;
+
+    /// @notice Constructs the factory.
+    /// @param _authorityFactory The `Authority` factory
+    /// @param _historyFactory The `History` factory
+    constructor(
+        IAuthorityFactory _authorityFactory,
+        IHistoryFactory _historyFactory
+    ) {
+        authorityFactory = _authorityFactory;
+        historyFactory = _historyFactory;
+
+        emit AuthorityHistoryPairFactoryCreated(
+            _authorityFactory,
+            _historyFactory
+        );
+    }
+
+    function getAuthorityFactory()
+        external
+        view
+        override
+        returns (IAuthorityFactory)
+    {
+        return authorityFactory;
+    }
+
+    function getHistoryFactory()
+        external
+        view
+        override
+        returns (IHistoryFactory)
+    {
+        return historyFactory;
+    }
+
+    function newAuthorityHistoryPair(
+        address _authorityOwner
+    ) external override returns (Authority authority_, History history_) {
+        authority_ = authorityFactory.newAuthority(address(this));
+        history_ = historyFactory.newHistory(address(authority_));
+
+        authority_.setHistory(history_);
+        authority_.transferOwnership(_authorityOwner);
+    }
+
+    function newAuthorityHistoryPair(
+        address _authorityOwner,
+        bytes32 _salt
+    ) external override returns (Authority authority_, History history_) {
+        _salt = calculateCompoundSalt(_authorityOwner, _salt);
+
+        authority_ = authorityFactory.newAuthority(address(this), _salt);
+        history_ = historyFactory.newHistory(address(authority_), _salt);
+
+        authority_.setHistory(history_);
+        authority_.transferOwnership(_authorityOwner);
+    }
+
+    function calculateAuthorityHistoryAddressPair(
+        address _authorityOwner,
+        bytes32 _salt
+    )
+        external
+        view
+        override
+        returns (address authorityAddress_, address historyAddress_)
+    {
+        _salt = calculateCompoundSalt(_authorityOwner, _salt);
+
+        authorityAddress_ = authorityFactory.calculateAuthorityAddress(
+            address(this),
+            _salt
+        );
+
+        historyAddress_ = historyFactory.calculateHistoryAddress(
+            authorityAddress_,
+            _salt
+        );
+    }
+
+    /// @notice Calculate the compound salt.
+    /// @param _authorityOwner authority owner
+    /// @param _salt salt
+    /// @return compound salt
+    /// @dev The purpose of calculating a compound salt is to add
+    /// entropy from the authority owner address to the address pair.
+    /// This is necessary because the authority owner address is not
+    /// passed to the `newAuthority` function, and, therefore, does not
+    /// influence the address of the authority contract.
```

Comment:
The purpose of adding entropy from the authority owner address is to prevent someone front-running the creation of Authority but with a different authority owner and occupying the to-be-deployed address. So we can summarize this paragraph into something like
"The purpose of calculating a compound salt is to prevent attackers front-running the creation of an Authority but with a different authority owner and occupying the to-be-deployed address."